### PR TITLE
feat: allow overriding preventDefault for modifier keys

### PR DESCRIPTION
**Pull request by jordanw66**
# Overview

In response to this issue: https://github.com/BuilderIO/qwik/issues/3871

The only place to preventDefault or not is in the loader, unfortunately this makes it impossible to preventDefault on anchors while also letting users use browser-native modifier keys (ctrl, shift, alt) to open new tabs, etc on these same links.

This impl. adds an attribute at the same level as `preventdefault:*` that allows overriding it for modifier keys. The reason for this is to allow devs to still preventDefault even on these modifier keys.

Given the widespread use of `preventdefault` and the niche use of modifier keys, I think keeping this check as false by default and opt-in only when needed makes more sense.

# What is it?

- [ ] Feature / enhancement
- [ ] Bug
- [ ] Docs / tests / types / typos

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

# Use cases and why

- 1. One use case
- 2. Another use case

# Checklist:

- [ ] My code follows the [developer guidelines of this project](https://github.com/BuilderIO/qwik/blob/main/CONTRIBUTING.md)
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] Added new tests to cover the fix / functionality

### CODE REVIEW

File: packages/qwik/src/qwikloader.ts
Comment (tzdesign) [inline packages/qwik/src/qwikloader.ts:65]: You always deliver @jordanw66. This dispatch is for all events right?

So this might be a good check and shortens the code a bit.
```suggestion
      if (
        !element.hasAttribute('allowmodifiers') ||
        !(ev instanceof MouseEvent && (ev.ctrlKey || ev.metaKey || ev.shiftKey || ev.altKey))
      ) {
        ev.preventDefault();
```

I'm maybe thinking to generic here and you might be right, this is enough.
As of right now, I have no clue how to document this that everyone gets it. Maybe this is not necessary as react-router-link works in the same way I guess.

**Review by tzdesign — CHANGES_REQUESTED** [review]

**Comment by shairez** [comment]
Thanks @jordanw66 !

Somewhat related to #4496 (as it's about adding a change to the qwikloader, which should be done on very rare ocassions IMO)

**Comment by tzdesign** [comment]
@shairez @jordanw66 this is even better. we can than fine-grain these options in Link or custom components.

It's not breaking if it's a union with boolean. @jordanw66 let me know if you need anything. I'm pretty free tomorrow.

File: packages/qwik/src/qwikloader.ts
Comment (tzdesign) [reply packages/qwik/src/qwikloader.ts:65]: @jordanw66 valid point, but what happens on keyup?

Speaking of performance, instanceof will be faster that the suggestion here: #4496.   
The alternative would be casting to unknown and then check with ```in```.

I guess you abort this one and create a new one? Let me know

**Comment by shairez** [comment]
> @shairez Interesting, didn't see that request. That will require a slightly larger increase to loader, do you guys know which direction you want to go in? I could take a stab at implementing that if you want. Could replace this PR with more functionality.

Thanks! yes this will resolve so many future issues as well 😅

The only concern I had is about how to convey that this is a syncornous function to the developer... but maybe if it's withouth the `$` then it's fine :)

Let's wait to hear what @manucorporat @adamdbradley @mhevery have to say about this before proceeding with a PR

**Review by manucorporat — CHANGES_REQUESTED** [review]
Gonna block this PR so it's not merged by mistake, i agree preventdefault need to be improved, but changing qwikloader is a very serious matter, and this would not solve all the problems.

Wonder if the solution is make preventdefault some sort of inlined function, but that would require some `new Function()` or `eval()`

**Comment by manucorporat** [comment]
The optimizer can already inline functions, but it's implicit there is not way to signal this
wonder if you can do this:

```
preventdefault:click={$$((event) => {

})}
```

$$ would be a new compiler API, like $, but syncronous, and you cant have imports!
the optimizer already knows to do that, but it's not public

**Comment by manucorporat** [comment]
Not a fan of `$$` and it would be very advanced API. Maybe:

`_inlined_fn_`

**Comment by tzdesign** [comment]
@manucorporat let us know if you know how to move forward. I think synchronous would be crazy in this async context.

I still can't get my head around, why we can't use event.preventDefault() in the Link-Component. You pass it through in dispatch, but I haven't read all of this and it ends up beeing any at some point.

I really need to read more about the qwik loader 🥲

**Comment by tzdesign** [comment]
@jordanw66 Thanks for the kick to my head,it works better now 😂

Your idea sounds good. Keeping track of them by referencing might be the best solution.

I have popcorn 🍿 now. Manu will find the right solution, he always does.

**Comment by tzdesign** [comment]
@manucorporat @jordanw66 is this still ongoing?

The browser's default behavior should actually be the default. I would propose `disallowModifiers` should be a prop.
For me even @jordanw66 PR would mean I have to change already over 100 links in our project.


I am thinking about writing a custom link component because of this behavior.
If you hit the meta keys to open in a new silent tab or window to copy the link faster or compare products in or ecommerce shop there is no way we will leave the link like this.

**Comment by DustinJSilk** [comment]
Hey guys, this issue has been stale for a long time, is it possible we could revive this and decide on a solution?
It's a poor experience to not be able to use default browser behaviour with links.

**Comment by wmertens** [comment]
@manucorporat So we want custom preventDefault filtering. Do we ever want any other sync event processing? What if we want to have the same custom preventDefault on several event types?

You proposed but did not like `preventdefault:click={$$((event) => {})}`, but I think we could special-case this? So just `preventDefault:whatever={(ev) => ...}`. The optimizer can emit the callback as an inline function and replace it with some sort of IQRL (inlined function index)?

**Comment by tzdesign** [comment]
@jordanw66 but is there a plan to implement it as default? Qwik is the only framework I know, overriding browser standards here.

**Comment by gabrielgrant** [comment]
@tzdesign this is tracked by #3871  right? looks like @mhevery self-assigned that, so guessing he's planning to make use of `sync()` to solve that once he's merged the `sync()` feature itself

**Comment by tzdesign** [comment]
> @tzdesign this is tracked by #3871  right? looks like @mhevery self-assigned that, so guessing he's planning to make use of `sync()` to solve that once he's merged the `sync()` feature itself

Oh gosh, I forgot to check my own issue 😅

Good to know. With @mhevery its in the best hands 🙌

**Pull request closed without merging**
